node/libs/schema/build.rs

```diff
@@ -1,159 +1,14 @@
-        check_canonical_message(&m).with_context(|| m.name().to_string())?;
-    }
-    Ok(())
-}
-
-/// Checks if message types in descriptor pool `d` support canonical encoding.
-fn check_canonical_pool(d: &prost_reflect::DescriptorPool) -> anyhow::Result<()> {
-    for f in d.files() {
-        check_canonical_file(&f).with_context(|| f.name().to_string())?;
-    }
-    Ok(())
-}
-
-fn main() -> anyhow::Result<()> {
-    // Prepare input and output root dirs.
-    let proto_include = PathBuf::from(env::var("CARGO_MANIFEST_DIR")?)
-        .canonicalize()?
-        .join("proto");
-    let proto_output = PathBuf::from(env::var("OUT_DIR").unwrap())
-        .canonicalize()?
-        .join("proto");
-    println!("cargo:rerun-if-changed={}", proto_include.to_str().unwrap());
-    let _ = fs::remove_dir_all(&proto_output);
-    fs::create_dir_all(&proto_output).unwrap();
-
-    // Find all proto files.
-    let mut proto_inputs = vec![];
-    traverse_files(proto_include.clone(), &mut |path| {
-        let Some(ext) = path.extension() else { return };
-        let Some(ext) = ext.to_str() else { return };
-        if ext != "proto" {
-            return;
-        };
-        proto_inputs.push(path);
-    })?;
-
-    // Generate protobuf code from schema (with reflection).
-    env::set_var("PROTOC", protoc_bin_vendored::protoc_bin_path().unwrap());
-    let mut config = prost_build::Config::new();
-    let descriptor_path = proto_output.join("descriptor.bin");
-    config.out_dir(&proto_output);
-    prost_reflect_build::Builder::new()
-        .file_descriptor_set_path(&descriptor_path)
-        .descriptor_pool("crate::proto::DESCRIPTOR_POOL")
-        .compile_protos_with_config(config, &proto_inputs, &[&proto_include])
-        .unwrap();
-    let descriptor = fs::read(descriptor_path)?;
-    let pool = prost_reflect::DescriptorPool::decode(descriptor.as_ref()).unwrap();
-
-    // Check that messages are compatible with `proto_fmt::canonical`.
-    check_canonical_pool(&pool)?;
-
-    // Generate mod file collecting all proto-generated code.
-    let mut m = Module::default();
-    for entry in fs::read_dir(&proto_output).unwrap() {
-        let entry = entry.unwrap();
-        let name = entry.file_name().into_string().unwrap();
-        let Some(name) = name.strip_suffix(".rs") else {
-            continue;
-        };
-        let name: Vec<_> = name.split('.').map(String::from).collect();
-        println!("name = {name:?}");
-        m.insert(&name, entry.path());
-    }
-    let file = m.generate();
-    let file = syn::parse_str(&file).unwrap();
-    fs::write(proto_output.join("mod.rs"), prettyplease::unparse(&file))?;
-    Ok(())
+//! Generates rust code from protobufs.
+fn main() {
+    zksync_protobuf::build::Config {
+        input_root: "proto".into(),
+        proto_root: "zksync/schema".into(),
+        dependencies: vec![(
+            "::zksync_protobuf::proto".into(),
+            &zksync_protobuf::proto::DESCRIPTOR,
```

Comment:
Having Protobuf dependencies as *build* dependencies is suboptimal since it requires the dependency to be fully built before building this crate even begins. I don't have immediate suggestions how to improve this though.

Reply:
I have some ideas, but they will be even more complex than what is here now. I was planning to first check the compilation time impact on zksync-era to see if it is worth the extra effort.

Reply:
Using exporting variables from the build scripts via fake linking could help; see [my comment above](https://github.com/matter-labs/era-consensus/pull/24#pullrequestreview-1717624969). OTOH, it could be useless, and it creates some complexity on its own (e.g., having to declare a fake `links` field in the crate manifest).